File: src/types/patient.ts
// ── Enums ──

export enum Gender {
  MALE = "MALE",
  FEMALE = "FEMALE",
  OTHER = "OTHER",
}

export enum AllergyStatus {
  CONFIRMED = "CONFIRMED",
  SUSPECTED = "SUSPECTED",
  DENIED = "DENIED",
}

// ── Sub-models ──

export interface ContactInfo {
  phone: string;
  address: string;
  emergency_contact_name: string;
  emergency_contact_relation: string;
  emergency_contact_phone?: string;
}

export interface Insurance {
  provider: string;
  plan_type: string;
}

export interface Allergy {
  allergen: string;
  reaction?: string;
  status?: AllergyStatus;
  recorded_at?: string;
}

export interface Medication {
  name: string;
  dosage: string;
  frequency: string;
  indication?: string;
  prescribed_at?: string;
}

export interface VitalSigns {
  blood_pressure_systolic?: number;
  blood_pressure_diastolic?: number;
  heart_rate?: number;
  temperature_fahrenheit?: number;
  bmi?: number;
  measurement_date?: string;
}

export interface ClinicalNote {
  note_type: string;
  date_of_service: string;
  provider_name: string;
  provider_credentials: string;
  subjective: string;
  objective: string;
  assessment: string;
  plan: string;
  chief_complaint?: string;
  vital_signs?: VitalSigns;
  follow_up_instructions?: string;
}

export interface Diagnosis {
  condition: string;
  icd_code?: string;
  date_diagnosed?: string;
  status?: "active" | "resolved" | "chronic";
}

export interface LabResult {
  test_name: string;
  result: string;
  unit: string;
  reference_range: string;
  flagged?: boolean;
  date_performed?: string;
}

export interface CompleteBloodCount extends LabResult {
  wbc?: number;
  rbc?: number;
  hemoglobin?: number;
  hematocrit?: number;
  platelets?: number;
}

export interface LipidPanel extends LabResult {
  total_cholesterol?: number;
  ldl_cholesterol?: number;
  hdl_cholesterol?: number;
  triglycerides?: number;
}

export interface ChemistryPanel extends LabResult {
  sodium?: number;
  potassium?: number;
  co2?: number;
  creatinine?: number;
  ast?: number;
  total_bilirubin?: number;
}

export interface ImagingStudy {
  study_type: string;
  body_part: string;
  findings: string;
  impression: string;
  date_performed: string;
  radiologist?: string;
}

export interface DiagnosticTest {
  test_type: string;
  date_performed: string;
  findings: string;
  interpretation: string;
  ordered_by?: string;
}

// ── Root model ──

export interface Patient {
  patient_id: string;
  medical_record_number: string;
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: Gender;
  contact_info: ContactInfo;
  insurance: Insurance;
  allergies: Allergy[];
  current_medications: Medication[];
  diagnoses: Diagnosis[];
  clinical_notes: ClinicalNote[];
  lab_results: LabResult[];
  imaging_studies: ImagingStudy[];
  diagnostic_tests: DiagnosticTest[];
  primary_care_physician?: string;
  hospital?: string;
  admission_date?: string;
  patient_signature?: string;
  signature_date?: string;
}

// ── Helper functions ──

export function getPatientAge(patient: Patient): number {
  return Math.floor(
    (Date.now() - new Date(patient.date_of_birth).getTime()) / 31557600000
  );
}

export function getActiveDiagnoses(patient: Patient): Diagnosis[] {
  return patient.diagnoses.filter((d) => d.status === "active");
}

export function getRecentLabs(patient: Patient, days = 30): LabResult[] {
  const cutoff = Date.now() - days * 86400000;
  return patient.lab_results.filter(
    (l) => l.date_performed && new Date(l.date_performed).getTime() >= cutoff
  );
}

File: src/components/ts_runtime_bundle/backend/models/patient.py
from dataclasses import dataclass, field
from datetime import date, datetime, timedelta
from typing import Optional, List

from .enums import Gender
from .contact import ContactInfo, Insurance
from .medical import Allergy, Medication
from .clinical import ClinicalNote, ImagingStudy, DiagnosticTest, Diagnosis
from .labs import LabResult


@dataclass
class Patient:
    """Complete patient medical record"""
    # Demographics
    patient_id: str
    medical_record_number: str
    first_name: str
    last_name: str
    date_of_birth: date
    gender: Gender

    # Contact and insurance
    contact_info: ContactInfo
    insurance: Insurance

    # Medical information
    allergies: List[Allergy] = field(default_factory=list)
    current_medications: List[Medication] = field(default_factory=list)
    diagnoses: List[Diagnosis] = field(default_factory=list)

    # Clinical records
    clinical_notes: List[ClinicalNote] = field(default_factory=list)
    lab_results: List[LabResult] = field(default_factory=list)
    imaging_studies: List[ImagingStudy] = field(default_factory=list)
    diagnostic_tests: List[DiagnosticTest] = field(default_factory=list)

    # Healthcare providers
    primary_care_physician: Optional[str] = None
    hospital: Optional[str] = None
    admission_date: Optional[date] = None

    # Administrative
    patient_signature: Optional[str] = None
    signature_date: Optional[datetime] = None

    def age(self) -> int:
        """Calculate patient age"""
        today = date.today()
        return today.year - self.date_of_birth.year - (
            (today.month, today.day) < (self.date_of_birth.month, self.date_of_birth.day)
        )

    def get_active_diagnoses(self) -> List[Diagnosis]:
        """Get list of active diagnoses"""
        return [d for d in self.diagnoses if d.status == "active"]

    def get_recent_labs(self, days: int = 30) -> List[LabResult]:
        """Get lab results from last N days"""
        cutoff = date.today() - timedelta(days=days)
        return [lab for lab in self.lab_results
                if lab.date_performed and lab.date_performed >= cutoff]
